### PR TITLE
Add a way to register listeners for app shutdown

### PR DESCRIPTION
**Pull request by jason-simmons**
Also add shutdown handlers that close Mojo handlers that were opened on behalf
of various services

### CODE REVIEW

File: packages/flutter/lib/src/services/fetch.dart [hunk 1]
```diff
@@ -39,7 +41,10 @@ Future<UrlResponse> fetch(UrlRequest request) async {
   UrlLoaderProxy loader = new UrlLoaderProxy.unbound();
   try {
     _networkService.ptr.createUrlLoader(loader);
+    ShutdownListener shutdownListener = () => loader.close();
+    lifecycle.addShutdownListener(shutdownListener);
     UrlResponse response = (await loader.ptr.start(request)).response;
+    lifecycle.removeShutdownListener(shutdownListener);
```
Comment (abarth) [inline packages/flutter/lib/src/services/fetch.dart]: In principle, wouldn't need need to do this for every await call on every mojo handle?  That seems crazy.

File: packages/flutter/lib/src/services/lifecycle.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import 'dart:ui' as ui;
+
+typedef void ShutdownListener();
+
+class _Lifecycle {
+  _Lifecycle() {
+    ui.lifecycleHooks.onShutdown = _handleShutdown;
+  }
+
+  final List<ShutdownListener> _shutdownListeners = new List<ShutdownListener>();
```
Comment (abarth) [inline packages/flutter/lib/src/services/lifecycle.dart]: Maybe this should be a LinkedHashSet?  What you've got here is n^2 in the number of in-flight mojo requests.

**Comment by abarth** [comment]
I don't think we should do this for the `await` case.  We need a better solution there.  For the statics, this makes much more sense.

**Comment by jason-simmons** [comment]
Abandoning this.  Registering callbacks on all Mojo handles to close them before shutdown feels like overkill.  Mojo will remove the warning about handles being closed by the finalizer.

Eventually we may need a more general solution to notify apps about activity lifecycle events.

**Pull request closed without merging**